keyboards/jones/v1/keymaps/via/keymap.c

```diff
@@ -0,0 +1,218 @@
+);
+
+const rgblight_segment_t PROGMEM my_lower_layer[] = RGBLIGHT_LAYER_SEGMENTS(
+    {JONES_LED_INDICATOR_INDEX , JONES_LED_INDICATOR_CHANGE_COUNT, HSV_GREEN - JONES_LED_DIMMER_LEVEL},
+    {JONES_LED_INDICATOR_INDEX + 2, JONES_LED_INDICATOR_CHANGE_COUNT, HSV_GREEN - JONES_LED_DIMMER_LEVEL}
+);
+
+const rgblight_segment_t PROGMEM my_raise_layer[] = RGBLIGHT_LAYER_SEGMENTS(
+    {JONES_LED_INDICATOR_INDEX , JONES_LED_INDICATOR_CHANGE_COUNT, HSV_CYAN - JONES_LED_DIMMER_LEVEL},
+    {JONES_LED_INDICATOR_INDEX + 2, JONES_LED_INDICATOR_CHANGE_COUNT, HSV_CYAN - JONES_LED_DIMMER_LEVEL}
+);
+
+const rgblight_segment_t PROGMEM my_adjust_layer[] = RGBLIGHT_LAYER_SEGMENTS(
+    {JONES_LED_INDICATOR_INDEX , JONES_LED_INDICATOR_CHANGE_COUNT, HSV_RED - JONES_LED_DIMMER_LEVEL},
+    {JONES_LED_INDICATOR_INDEX + 2, JONES_LED_INDICATOR_CHANGE_COUNT, HSV_RED - JONES_LED_DIMMER_LEVEL}
+);
+
+// Define the array of layers. Later layers take precedence
+const rgblight_segment_t* const PROGMEM my_rgb_layers[] = RGBLIGHT_LAYERS_LIST(
+    my_base1_layer,
+    my_base2_layer,
+    my_base3_layer,
+    my_caps_layer,
+    my_lower_layer,
+    my_raise_layer,
+    my_adjust_layer
+);
+
+// Enabling and disabling lighting layers
+layer_state_t layer_state_set_user(layer_state_t state) {
+    rgblight_set_layer_state(4, layer_state_cmp(state, _LOWER));
+    rgblight_set_layer_state(5, layer_state_cmp(state, _RAISE));
+    rgblight_set_layer_state(6, layer_state_cmp(state, _ADJUST));
+
+    return state;
+}
+
+// Enabling and disabling lighting layers for default layer
+layer_state_t default_layer_state_set_user(layer_state_t state) {
+    rgblight_set_layer_state(0, layer_state_cmp(state, _BASE1));
+    rgblight_set_layer_state(1, layer_state_cmp(state, _BASE2));
+    rgblight_set_layer_state(2, layer_state_cmp(state, _BASE3));
+
+    return state;
+}
+
+bool led_update_user(led_t led_state) {
+    rgblight_set_layer_state(3, IS_HOST_LED_ON(USB_LED_CAPS_LOCK));
+
+    return true;
+}
+#endif
+
+
+//------------------------------------------------------------------------------
+// Rotary Encoder
+//------------------------------------------------------------------------------
+bool encoder_update_user(uint8_t index, bool clockwise) {
+    keypos_t key;
+
+    // Both encoders trigger specific key matrix position when turn knob clockwise or counterclockwise.
+
+    if (index == 0) { // First encoder, Left side : k85, k86
+        if (clockwise) {
+            key.row = 8;
+            key.col = 6;
+        } else {
+            key.row = 8;
+            key.col = 5;
+        }
+    } else if (index == 1) { // Second encoder, Right side : k95, k96
+        if (clockwise) {
+            key.row = 9;
+            key.col = 6;
+        } else {
+            key.row = 9;
+            key.col = 5;
+        }
+    }
+
+    uint8_t  layer   = layer_switch_get_layer(key);
+    uint16_t keycode = keymap_key_to_keycode(layer, key);
+
+    tap_code16(keycode);
```

Comment:
action_exec is a better way to handle this, because it will support ANY keycode, including custom keycodes.  

`tap_code16` will only support basic keycodes, with or without mods, but nothing else. 

You can see an example of this here: 
https://github.com/qmk/qmk_firmware/blob/3496513865b0b1ac1b4c3771c64581e626428c4d/keyboards/work_louder/encoder_actions.c#L22-L68

Reply:
Implemented better way to handle keycode.

I could not write better code than your example.
So, what I could do is just copy and modify values...

Reply:
That's absolutely fine!